### ISSUE TITLE
generated config.sh messes up LD_LIBRARY_PATH

### ISSUE THREAD

**Issue by spflueger** [FairRootGroup/FairRoot#796]
This issue is related to #723 
When I load a different compiler than the standard one, the corresponding shared libraries directories are not put into the LD_LIBRARY_PATH env in config.sh generated by FairRoot. Somehow the LD_LIBRARY_PATH when cmake is executed (for pandaroot) removes the gcc lib directories from the list. And since also no appending of the LD_LIBRARY_PATH is done, the config.sh script corrupts my LD_LIBRARY_PATH.

**Comment by dennisklein** [comment]
There is a solution in `dev` for this. It allows you add `-p`/`-a` to the `config.sh` call which has the effect of prepending/appending to `LD_LIBRARY_PATH` instead of overwriting it. Maybe you can cherry-pick 060cf2bd0245ce444d1506 for now or @fuhlig1 has time to backport it for the 17.10 release. I myself won't have time to do it until third week of July, I am afraid.

**Comment by spflueger** [comment]
Hmm ok, I think I did not explain my problem very nicely. I understand that Fairroot wants a clean LD_LIBRARY_PATH that it cant get messy with different versions of dependencies etc (a bit like a virtual env). But why are the compiler standard libraries, which are needed to run any fairroot macro or application removed from LD_LIBRARY_PATH. Basically I cant run any Fairroot code unless I manually add my compiler library directory to the LD_LIBRARY_PATH.

This is my LD_LIB_PATH before i setup Pandaroot

> /cluster/easybuild/broadwell/software/lang/Python/3.6.3-foss-2017a/lib/python3.6/site-packages/numpy-1.13.3-py3.6-linux-x86_64.egg/numpy/core/lib:/cluster/easybuild/broadwell/software/lang/Python/3.6.3-foss-2017a/lib:/cluster/easybuild/broadwell/software/numlib/ScaLAPACK/2.0.2-gompi-2017a-OpenBLAS-0.2.19-LAPACK-3.7.0/lib:/cluster/easybuild/broadwell/software/numlib/FFTW/3.3.6-gompi-2017a/lib:/cluster/easybuild/broadwell/software/numlib/OpenBLAS/0.2.19-GCC-6.3.0-LAPACK-3.7.0/lib:/cluster/easybuild/broadwell/software/mpi/OpenMPI/2.0.2-GCC-6.3.0/lib:**/cluster/easybuild/broadwell/software/compiler/GCCcore/8.1.0/lib/gcc/x86_64-pc-linux-gnu/8.1.0:/cluster/easybuild/broadwell/software/compiler/GCCcore/8.1.0/lib64:/cluster/easybuild/broadwell/software/compiler/GCCcore/8.1.0/lib**

and this is what the generated config.sh will make out of it:

> /cluster/easybuild/broadwell/software/lang/Python/3.6.3-foss-2017a/lib/python3.6/site-packages/numpy-1.13.3-py3.6-linux-x86_64.egg/numpy/core/lib:/cluster/easybuild/broadwell/software/lang/Python/3.6.3-foss-2017a/lib:/cluster/easybuild/broadwell/software/numlib/ScaLAPACK/2.0.2-gompi-2017a-OpenBLAS-0.2.19-LAPACK-3.7.0/lib:/cluster/easybuild/broadwell/software/numlib/FFTW/3.3.6-gompi-2017a/lib:/cluster/easybuild/broadwell/software/numlib/OpenBLAS/0.2.19-GCC-6.3.0-LAPACK-3.7.0/lib:/cluster/easybuild/broadwell/software/mpi/OpenMPI/2.0.2-GCC-6.3.0/lib:/home/stpflueg/PandaRoot/build/lib:/home/stpflueg/FairRootInstall/lib:/cluster/him/fairsoft/gcc_8.1.0/oct17p1/lib/root:/cluster/him/fairsoft/gcc_8.1.0/oct17p1/lib

I highlighted the important cut out part, which is missing in config.sh. This never happened to me before. I have the feeling it has something to do with my computers configuration. I will try to find out why tomorrow.

**Comment by dennisklein** [comment]
>  I understand that Fairroot wants a clean LD_LIBRARY_PATH that it cant get messy with different versions of dependencies etc (a bit like a virtual env)

I do not to be honest. `LD_LIBRARY_PATH` is a shared facility and any virtualenv-like tool I know retains the content of this variable (and also `PATH`). But anyways, the behaviour of `config.sh` was, is and will be to take ownership over these variables. As you can read in #723, we added two additional options to `config.sh` - namely `-p` (prepend) and `-a` (append). So, anyone, who likes to mix the FairRoot environment with his other environments, can choose to do so. These options are not yet in the 17.10c release of FairRoot.

Capturing your environment at build time and burning it into `config.sh` is IMHO not a good solution, but even that we support with the CMake build switch `-DUSE_PATH_INFO=TRUE`. Though I have never tried it, not sure, if your described behaviour is due to a bug with that option? Are you saying that?

If you ask me, I recommend to setup your environment as usual and then call `config.sh -p` to add the FairRoot environment. If you used a custom compiler environment for building, you would set it up the same way before calling `config.sh -p`.

FairRoot will definitely never start to try to understand your environment at build time and be smart about retaining one path over the other, because it may contain a standard library.

To summarize, you got four options, which should cover all meaningful scenarios:

| Option | Effect | Version |
| --- | --- | --- |
| `-DUSE_PATH_INFO=TRUE` | Burns your build time env into `config.sh` | Supported in **17.10c** |
| `source config.sh` | Overwrites any existing env | Supported in **17.10c** |
| `source config.sh -p` | Prepends the FairRoot env to the current | see  060cf2bd0245ce444d15069989265ed95c1acb44 |
| `source config.sh -a` | Appends the FairRoot env to the current | see 060cf2bd0245ce444d15069989265ed95c1acb44 |

**Comment by spflueger** [comment]
It seems like a bug to me. I have worked with fairroot and pandaroot for years now. And after the installation the only thing I had to do was run the config.sh (which was generated by fairroot+pandaroot). Then I was able to run the code. So that is your option 2. However this does not work anymore as the env is messed up (see the comment from before). The config generator just takes my compiler settings and throws them away.

**Comment by fuhlig1** [comment]
Did you use also before a „non-system“ compiler or the compiler which was installed with the system?

Ciao

Florian

**Comment by spflueger** [comment]
I tried to set it up freshly and I only used a non-system compiler gcc 8.1

**Comment by fuhlig1** [comment]
Sorry but my last mail was not properly phrased.

> I have worked with fairroot and pandaroot for years now. And after the installation the only thing I had to do was run the config.sh (which was generated by fairroot+pandaroot). Then I was able to run the code. So that is your option 2. However this does not work anymore as the env is messed up (see the comment from before). The config generator just takes my compiler settings and throws them away.


You claim that you never had problems with the installation procedure. What I want to know is if you have used an external compiler before when everything was working. The default behaviour when generating the config file is to throw away your PATH and LD_LIBRARY_PATH variables. Please try to run cmake with the parameter -DUSE_PATH_INFO=TRUE which will keep this information and should create a proper config.sh script.

Ciao

Florian

**Comment by spflueger** [comment]
Yeah I have used external compilers before many times, and I believe calling config.sh always worked. Maybe I'm wrong and I always prepended my compiler settings to the env again after the sourcing of the config.sh.... I'm just debugging this issue

**Comment by spflueger** [comment]
@fuhlig1 With -DUSE_PATH_INFO=TRUE everything works fine. Was that always the default behaviour to throw away PATH and LD_LIBRARY_PATH? If so, then I must just remember wrong and I always fixed the env after the sourcing of config.sh. Not very convenient though. Is it possible to make this appending (or prepending) flag default?
